runtime/runenv_events.go

```diff
@@ -1,61 +1,166 @@
 package runtime
 
 import (
+	"context"
 	"fmt"
 	"runtime/debug"
 
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
-type (
-	EventType    string
-	EventOutcome string
-)
+type Event struct {
+	*StartEvent      `json:"start_event,omitempty"`
+	*MessageEvent    `json:"message_event,omitempty"`
+	*SuccessEvent    `json:"success_event,omitempty"`
+	*FailureEvent    `json:"failure_event,omitempty"`
+	*CrashEvent      `json:"crash_event,omitempty"`
+	*StageStartEvent `json:"stage_start_event,omitempty"`
+	*StageEndEvent   `json:"stage_end_event,omitempty"`
+}
 
-const (
-	EventTypeStart   = EventType("start")
-	EventTypeMessage = EventType("message")
-	EventTypeFinish  = EventType("finish")
+func (e *Event) Type() string {
+	switch {
+	case e.StartEvent != nil:
+		return e.StartEvent.Type()
+	case e.MessageEvent != nil:
+		return e.MessageEvent.Type()
+	case e.SuccessEvent != nil:
+		return e.SuccessEvent.Type()
+	case e.FailureEvent != nil:
+		return e.FailureEvent.Type()
+	case e.CrashEvent != nil:
+		return e.CrashEvent.Type()
+	case e.StageStartEvent != nil:
+		return e.StageStartEvent.Type()
+	case e.StageEndEvent != nil:
+		return e.StageEndEvent.Type()
+	default:
+		panic("no such event")
+	}
+}
 
-	EventOutcomeOK      = EventOutcome("ok")
-	EventOutcomeFailed  = EventOutcome("failed")
-	EventOutcomeCrashed = EventOutcome("crashed")
-)
+type StartEvent struct {
+	Runenv *RunParams `json:"runenv"`
+}
 
-type Event struct {
-	Type       EventType    `json:"type"`
-	Outcome    EventOutcome `json:"outcome,omitempty"`
-	Error      string       `json:"error,omitempty"`
-	Stacktrace string       `json:"stacktrace,omitempty"`
-	Message    string       `json:"message,omitempty"`
-	Runenv     *RunParams   `json:"runenv,omitempty"`
+func (StartEvent) Type() string {
+	return "StartEvent"
```

Comment:
Should we use snake case for consistent casing?

Reply:
Fix this in the next PR already :)